storage/connectors/redis/src/main/java/feast/storage/connectors/redis/writer/RedisCustomIO.java

```diff
@@ -242,7 +266,22 @@ public void processElement(ProcessContext context) {
                                   .thenAccept(
                                       currentValue -> {
                                         FeatureRow newRow = entry.getValue();
+                                        FeatureSetSpec spec =
+                                            latestSpecs.get(newRow.getFeatureSet());
                                         if (rowShouldBeWritten(newRow, currentValue)) {
+                                          if (redisIngestionClient.getEnableRedisTtl()) {
```

Comment:
I don't understand why we need here `setRedisTtl` check. and how it affect our decision write row to redis or not?

Reply:
Does it mean that feature sets without max-age won't be written to redis?

Reply:
I think this part should be to line 298 `if (keysWithTtl.containsKey(key))`, bc here it clearly creates a bug:
1. this lambda function is being called from multiple threads, and `keysWithTtl` is not thread safe
2. if row don't pass `setRedisTtl` it's not being written at all (unit test needed)

Reply:
Ok, I get it. In this part we checking whether `eventTimestamp + ttl` < currentTime. Thus if row is outdated already, we won't write it at all. But implementation is very confusing and `setRedisTtl` interface and name as well.
Also this function is being called from many threads. Are you sure `keysWithTtl ` is thread safe?

Reply:
OK, to prevent method that modifies its arguments I suggest next
```
List<Triple<RedisKey, FeatureRow, lon>> rowsWithTtl = 
   deduplicatedRows.entrySet.stream()
        .map((key, row) -> 
           Triple(key, row, calculateRowTtl(row, latestSpecs.get(row.getFeatureSet()), redisIngestionClient.getMaxRedisTtlJitterSeconds(), redisIngestionClient.getMaxRedisTtlSeconds())
        .filter(triple -> triple._3 > 0).collect(...)

....
long calculateRowTtl(FeatureRow, FeatureSetSpec spec, int maxRedisTtlJitterSeconds, long maxRedisTtl)

Reply:
```
skippedFeatureRow = deduplicatedRows.filter(not in rowsWithTtl)
```

Reply:
Got it, thank you for the feedback. I'll address this in the next couple of days.

Reply:
@pyalex I made a few changes to avoid the call by reference issues you brought up. You're correct that we wanted to avoid writes if they are unnecessary (i.e. if the TTL <= 0). We thought it would be helpful to log when this happens, which is why we're keeping track of the rows that are skipped for this reason. We also need to distinguish where we are not using a TTL (either because TTL has not been enabled or the feature row does not have a max age we can use).

I created a `calculateRowTtl` function as you suggested that returns the row TTL or null if TTL should not be used. If the TTL is not a positive number, then the caller is responsible for skipping the write. Hopefully this makes sense!

Reply:
@radvin I agree that we need to log skipped row. Visibility is always good. My question is: do we really need to read from redis before skip row with `ttl < 0`?

My feeling is that independently of what stored in redis we won't want to write outdated (`ttl < 0`) value. This means that we can filtered out those rows earlier and make read batch smaller.

Also I still concerned about this line
```keysWithTtl.put(entry.getLeft(), ttl);```
This is inside lambda function, that is being called by lettuce in potentially multi-threaded executor and `keysWithTtl` is not thread-afe.

Reply:
@pyalex I made these changes as suggested:

1. We're no longer reading from redis if the TTL is calculated to be <= 0. This means that these rows will be ignored regardless of what is already in the database.
2. I made the `keysWithTtl` hashmap thread safe by moving writes outside the lambda function.